### PR TITLE
Define how to show and apply modifiers

### LINKED ISSUE #60: Define "last one wins" approach when Basic Card modifiers are equivalent
My understanding of the consensus that has emerged through discussion of PR API issue 684 [1] is that each specification should define its own approach to resolving duplicate instances of modifiers, and that payment method specifications should favor a "last one wins" approach. 

As I understand it, this is the approach taken by current implementations of Basic Card. 

I am raising an issue here so that we can close issue 684 on PR API more quickly (with an informative note in that specification). @marcoscaceres has indicated he would work on the text in Basic Card to capture the "last one wins" approach. However, we propose to wait until resolving issue #59 because the algorithm will be simpler if we drop supportedTypes.

[1] https://github.com/w3c/payment-request/issues/684#issuecomment-405612400

### PR DESCRIPTION
**Pull request by marcoscaceres**
closes #60 

The following tasks have been completed:

 * [x] Confirmed there are no ReSpec errors/warnings.
 * [x] [Added Web platform tests](
https://github.com/web-platform-tests/wpt/pull/15973) 
 * [ ] added MDN Docs (link)

Implementation commitments:

 * [ ] [Chrome](https://crbug.com/944817)
 * [ ] [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1538067)
 * [ ] Edge (public signal)

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/payment-method-basic-card/pull/71.html" title="Last updated on Mar 23, 2019, 5:10 AM UTC (e065781)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/payment-method-basic-card/71/29e3f10...e065781.html" title="Last updated on Mar 23, 2019, 5:10 AM UTC (e065781)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -458,6 +467,137 @@ <h4>
...
+          </pre>
+        </div>
+        <ol class="algorithm">
+          <li>For each <a data-cite="infra#pair">pair</a> (<a>DOMString</a>
+          <var>id</var> / string <var>data</var>) of
+          <var>request</var>.<a data-cite=
+          "payment-request/#dfn-serializedmethoddata">[[\serializedMethodData]]</a>
+          in reverse order:
+            <ol>
+              <li>If <var>id</var> is not "<a>basic-card</a>", continue.
```
Comment (marcoscaceres) [inline index.html]: Note to self: deal with null case. 

```
            // ✅ This second one gets used!
            {
              supportedMethods: "basic-card",
              data: {
                supportedNetworks: ["visa"],
              },
            },
            // null case gets ignored. 
            {
              supportedMethods: "basic-card",
            },
```

File: index.html [hunk 2]
```diff
@@ -458,6 +467,140 @@ <h4>
...
+          "payment-request#dom-paymentrequest-show"><code>show()</code></a>
+          method. The steps take a <a>PaymentRequest</a> <var>request</var>.
+        </p>
+        <div class="note">
+          <p>
+            In the case where a payment request is shown with multiple
+            applicable <a>PaymentMethodData</a>s, this algorithm selects the
+            last <a data-cite=
+            "payment-request/#dfn-serializedmethoddata">[[\serializedMethodData]]</a>
+            whose PMI is "<code>basic-card</code>". (i.e., "the last one
```
Comment (rsolomakhin) [inline index.html]:
```suggestion
            whose PMI is "<code>basic-card</code>", (i.e., "the last one
```

**Review by rsolomakhin — APPROVED** [review]
Algorithms seem sensible. Thank you.

**Comment by marcoscaceres** [comment]
@rsolomakhin thanks for having a look. 

@danyao, hopefully less terrible than last time 😅.

File: index.html [hunk 3]
```diff
@@ -458,6 +467,142 @@ <h4>
...
+              </li>
+              <li>Apply <var>restrictions</var>'s <a>supportedNetworks</a>
+              member to constrain the card type inputs of the payment UI.
+              </li>
+              <li>Break.
+              </li>
+            </ol>
+          </li>
+          <li>If an payment instrument is selected by default, <a>apply the
+          modifiers</a> using <var>request</var><a data-cite=
```
Comment (danyao) [inline index.html]:
```suggestion
          modifiers</a> using <var>request</var>.<a data-cite=
```

File: index.html [hunk 4]
```diff
@@ -458,6 +467,142 @@ <h4>
...
+          "payment-request/#dfn-serializedmethoddata">[[\serializedMethodData]]</a>
+          in reverse order:
+            <ol>
+              <li>If <var>id</var> is not "<a>basic-card</a>", continue.
+              </li>
+              <li>Let <var data-type="object">object</var> be result of
+              <a data-cite="ecma-262#sec-json.parse">JSON-parsing</a>
+              <var>data</var>.
+              </li>
+              <li>If <var>object</var> is <code>null</code>, continue.
```
Comment (danyao) [inline index.html]: Can `object` also be `null` if the original `PaymentMethodData` didn't contain a `data` field? Should the absence of `data` field also be considered "all networks are supported"?

File: index.html [hunk 5]
```diff
@@ -458,6 +467,142 @@ <h4>
...
+              </li>
+              <li>Let <var>bcRequest</var> be the result of converting
+              <var>object</var> to <a>BasicCardRequest</a>.
+              </li>
+              <li>If <var>bcRequest</var>'s <a>supportedNetworks</a> is zero
+              length, or <a>supportedNetworks</a> includes
+              <var>networkIdentifier</var>:
+                <ol>
+                  <li>Let <var>modifier</var> be the PaymentMethodModifier from
+                  <var>request</var>.[[\details].modifiers at <var>index</var>.
```
Comment (danyao) [inline index.html]:
```suggestion
                  <var>request</var>.[[details]].modifiers at <var>index</var>.
```

File: index.html (hunk 4 above)
Comment (marcoscaceres) [reply index.html]: > Can object also be null if the original PaymentMethodData didn't contain a data field?

Yes, the PR API substitutes missing with null. 

> Should the absence of data field also be considered "all networks are supported"?

I don’t feel strongly about it. From way back, we’d agreed that only “[]” would mean “all”. Thus I’m inclined to treat no data as “ignore”.

File: index.html (hunk 5 above)
Comment (marcoscaceres) [reply index.html]: This one is correct. The “\” implies “internal slot” rather than citation.

Comment (danyao) [reply index.html]: Hmm elsewhere I see `request.[[serializedMethodData]]` being used to reference the `[[serializedMethodData]]` internal slot of `request`. Perhaps I'm missing something.

**Comment by danyao** [comment]
LGTM % small question

**Review by danyao — APPROVED** [review]
Sorry for clicking on the wrong button earlier! LGTM % a small question.

File: index.html (hunk 5 above)
Comment (marcoscaceres) [reply index.html]: Yes, you are absolutely correct. I need to go through and fix all those and a bunch of links. Just wanted to get this up for a sanity check. Will fix it up tomorrow.

Comment (danyao) [reply index.html]: SG. Thanks!

**Comment by marcoscaceres** [comment]
Thanks @danyao and @rsolomakhin for the speedy reviews. I’ll see about putting some tests together too. Ping you over on the WPT repo.

**Comment by marcoscaceres** [comment]
@danyao, @rsolomakhin... been thinking a lot about @danyao's comment about treating missing as meaning "all networks". After spec'ing this out and thinking of possible test cases, I think what @danyao was suggesting makes a lot of sense. 

Consider the following cases. The first will be quite common, and I think browsers already support this:

```JS
// Semantically, these are all the same:  
[{ paymentMethod: "basic-card" }]
[{ paymentMethod: "basic-card", data: {} }]
[{ paymentMethod: "basic-card", data: { supportedNetworks: [] }}]
```

So to get that to mean "all the networks" we need to give `supportedNetworks` a default value` = []` in the IDL. 

Consider in particular the second case (`data: {}`)... passing that today would mean "I don't support any cards at all", which doesn't make sense for a "basic-card" request, right? Why say, "I support basic card... but I don't support any networks, lol."   

Additionally, for the case where a payment request is constructed with `[{ paymentMethod: "basic-card"}]`, we need to pretend that MethodData `.data` member is defaulting to a newly created `BasicCardPayment` dictionary. I've added prose to deal with that (the `restrictions` variable).

The same rules then need to apply to modifiers too. Consider:

```JS
// What does this mean when it's missing .data? 
const modifiers = [
  {
    supportedMethods: "basic-card",
    total: {
      label: "Total due",
      amount: { currency: "USD", value: "68.00" },
    },
  },
];
```

If we follow the same rules as I described above, then missing `.data` implies a `BasicCardRequest` whose `supportedNetworks = []`.

**Review by rsolomakhin — APPROVED** [review]
I agree with you, @marcoscaceres, that `modifiers: {supportedMethods: 'basic-card'}` should mean all networks.

File: index.html [hunk 6]
```diff
@@ -458,6 +469,160 @@ <h4>
           </dl>
         </section>
       </section>
+      <section>
+        <h2>
+          Showing the payment handler
```
Comment (ianbjacobs) [inline index.html]:
```suggestion
          Selecting the payment handler
```

Comment (ianbjacobs) [reply index.html]: I advise against the word "show" this is not about User Agent display.

File: index.html [hunk 7]
```diff
@@ -458,6 +469,160 @@ <h4>
...
+        </h2>
+        <div class="note">
+          <p>
+            In the case where a payment request is shown with multiple
```
Comment (ianbjacobs) [inline index.html]:
```suggestion
            In the case where a payment request is made with multiple
```

File: index.html [hunk 8]
```diff
@@ -458,6 +469,160 @@ <h4>
...
+            applicable <a>PaymentMethodData</a>s, this algorithm selects the
+            last applicable <a data-cite=
+            "payment-request#dfn-serializedmethoddata">[[\serializedMethodData]]</a>
+            whose PMI is "<code>basic-card</code>". (i.e., "the last one
+            wins").
+          </p>
+          <pre class="example" title="Last one wins">
+              const methodData = [
+              // This one is ignored.
```
Comment (ianbjacobs) [inline index.html]:
```suggestion
              // This one matches, but is ignored since one the follows also matches.
```

File: index.html [hunk 9]
```diff
@@ -458,6 +469,160 @@ <h4>
...
+                supportedMethods: "https://example.com/bobpay",
+              },
+            ];
+
+            // See above for which methodData gets picked.
+            await new PaymentRequest(methodData, details, options).show();
+            </pre>
+        </div>
+        <p>
+          The <dfn>steps for showing the payment handler</dfn> are given by the
```
Comment (ianbjacobs) [inline index.html]:
```suggestion
          The <dfn>steps for selecting the payment handler</dfn> are given by the
```

File: index.html [hunk 10]
```diff
@@ -458,6 +469,160 @@ <h4>
...
+          Applying the modifiers
+        </h2>
+        <p>
+          The steps to <dfn>apply the modifiers</dfn> are given by the
+          following algorithm. It takes a <a>PaymentRequest</a> <var data-type=
+          "PaymentRequest">request</var> and a string <var data-type=
+          "string">networkIdentifier</var> that identifies a <a>network</a>:
+        </p>
+        <ol class="algorithm">
+          <li>Let <var data-type="number">index</var> the length of
```
Comment (ianbjacobs) [inline index.html]:
```suggestion
          <li>Let <var data-type="number">index</var> be the length of
```

**Review by ianbjacobs — COMMENTED** [review]
@marcoscaceres, thank you. See a few things inline. I read the text but would prefer that an implementer approve the algorithm.

Ian

File: index.html (hunk 6 above)
Comment (marcoscaceres) [reply index.html]: Debatable because we literally get here from `.show()`, but yeah.

**Comment by marcoscaceres** [comment]
Thanks @ianbjacobs - good suggestions!

File: index.html (hunk 6 above)
Comment (ianbjacobs) [reply index.html]: I think that the selection algorithm may be applicable for more scenarios than just show().

Comment (marcoscaceres) [reply index.html]: Potentially, yes.

**Comment by danyao** [comment]
@marcoscaceres Treating `{supportedMethods: 'basic-card'}` consistently as supporting all networks in both payment method and modifiers sounds good. 👍

**Comment by marcoscaceres** [comment]
@danyao or @rsolomakhin, when possible, could you please provide a link to a Chrome bug?

**Comment by rsolomakhin** [comment]
> provide a link to a Chrome bug?

Done. Linked to https://crbug.com/944817 in the description.

**Comment by marcoscaceres** [comment]
Thanks @rsolomakhin. I'll patch Firefox.

**Pull request merged**